crates/rpc/rpc/src/eth/helpers/pending_block.rs

```diff
@@ -20,14 +32,90 @@ where
-                          + ChainSpecProvider<ChainSpec: EthChainSpec + EthereumHardforks>
+                          + ChainSpecProvider<ChainSpec = ChainSpec>
                           + StateProviderFactory,
             Pool: TransactionPool,
             Evm: ConfigureEvm<Header = Header>,
         >,
+    ChainSpec: EthChainSpec + EthereumHardforks,
 {
     #[inline]
     fn pending_block(&self) -> &tokio::sync::Mutex<Option<PendingBlock>> {
         self.inner.pending_block()
     }
+
+    /// Builds a pending block using the configured provider and pool.
+    ///
+    /// If the origin is the actual pending block, the block is built with withdrawals.
+    ///
+    /// After Cancun, if the origin is the actual pending block, the block includes the EIP-4788 pre
+    /// block contract call using the parent beacon block root received from the CL.
+    fn build_block(
+        &self,
+        env: PendingBlockEnv,
+    ) -> Result<(SealedBlockWithSenders, Vec<Receipt>), Self::Error>
+    where
+        EthApiError: From<ProviderError>,
+    {
+        let PendingBlockEnv { cfg: _, block_env, origin } = env;
+
+        // Prepare payload attributes
+        let payload_attributes = PayloadAttributes {
+            parent_beacon_block_root: origin.header().parent_beacon_block_root,
+            prev_randao: block_env.prevrandao.unwrap_or_default(),
+            timestamp: block_env.timestamp.to::<u64>(),
+            suggested_fee_recipient: block_env.coinbase,
+            withdrawals: self
+                .provider()
+                .chain_spec()
+                .is_shanghai_active_at_timestamp(block_env.timestamp.to::<u64>())
+                .then(Default::default),
+        };
+
+        // Create the payload config
+        let payload_config = PayloadConfig::new(
+            Arc::new(origin.header().clone()),
+            Bytes::default(),
+            reth_payload_builder::EthPayloadBuilderAttributes::try_new(
+                origin.header().hash(),
+                payload_attributes,
+                EngineApiMessageVersion::default() as u8, // Engine API message version
+            )
+            .unwrap(), // Unwrap the Result to get EthPayloadBuilderAttributes
+        );
+
+        // Build arguments
+        let args = BuildArguments::new(
+            self.provider().clone(),
+            self.pool().clone(),
+            CachedReads::default(),
+            payload_config,
+            Cancelled::default(), // Or provide a cancellation token if necessary
+            None,                 // Optional best payload
+        );
+
+        // Instantiate the payload builder locally
+        let evm_config = self.evm_config().clone();
+        let payload_builder = EthereumPayloadBuilder::new(evm_config);
+
+        // Build the payload
+        match payload_builder.try_build(args).unwrap() {
+            BuildOutcome::Better { payload, .. } => {
+                let sealed_block = payload.block().clone();
```

Comment:
can't this use the configured payload builder?

https://github.com/paradigmxyz/reth/blob/33730536f5e5de1403f2deb3c7dbe551879b9724/crates/rpc/rpc-eth-api/src/node.rs#L34-L35